network/resolvconf.go

```diff
@@ -0,0 +1,78 @@
+package network
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"syscall"
+)
+
+const resolv = "/etc/resolv.conf"
+
+func InstallResolvconf(name, containerDNS string) error {
+	containerDNSResolv, err := touchXDGFile("dagger/net/" + name + "/resolv.conf")
+	if err != nil {
+		return err
+	}
+
+	// create the resolv.conf for the container namespace by swapping out the
+	// nameservers from the original, keeping any options and search domains
+	if err := replaceNameservers(containerDNS, containerDNSResolv); err != nil {
+		return fmt.Errorf("replace nameservers: %w", err)
+	}
+
+	upstreamResolv, err := touchXDGFile("dagger/net/" + name + "/resolv.conf.upstream")
+	if err != nil {
+		return err
+	}
+
+	// preserve original resolv.conf at upstream path
+	//
+	// if resolv.conf is bind mounted, its source will be bind mounted here
+	if err := syscall.Mount(resolv, upstreamResolv, "", syscall.MS_BIND, ""); err != nil {
+		return fmt.Errorf("remount /etc/resolv.conf to upstream alias: %w", err)
+	}
+
+	// unmount target resolv.conf so we can replace it
+	if err := syscall.Unmount(resolv, 0); err != nil && !errors.Is(err, syscall.EINVAL) {
+		return fmt.Errorf("unmount /etc/resolv.conf: %w", err)
+	}
```

Comment:
Good point - I had it this way because my original approach involved unmounting and writing to the file directly, but since we're mounting over it instead we can just let it stack. :+1: